[Mono][jit] Emit GC transitions for "calli unmanaged" instructions

lambdageek opened the pull request:
- Add `mono_marshal_get_native_func_wrapper_indirect` 
   This will be used to add a wrapper around `calli <sig>` indirect calls where `sig` has an unmanaged calling convention.

   We need to do a GC transition before calling an unmanaged function from managed. So this wrapper does it.

   This is reusing much of the code implemented for `mono_marshal_get_native_func_wrapper_aot` which is used for delegates.

- Treat CallConv bit 0x09 as `MONO_CALL_UNMANAGED_MD` 
   This is the C#9 function pointers "unmanaged"+ext calling convention.
   Additional calling convention details are encoded in the modopts of the return type.

   This PR doesn't handle the modopts yet

- Rewrite `mono_marshal_emit_native_wrapper` to take a new `MonoNativeWrapperFlags` argument instead of a collection of bools. Add a new `EMIT_NATIVE_WRAPPER_FUNC_PARAM_UNBOXED` value.  Add support to emit a wrapper that uses an unboxed function pointer as the indirect arg.

Related to https://github.com/dotnet/runtime/issues/38480

lambdageek committed:
[metadata] Treat CallConv bit 0x09 as MONO_CALL_UNMANAGED_MD

This is the C#9 function pointers "unmanaged"+ext calling convention.
Additional calling convention details are encoded in the modopts of the return
type.

This PR doesn't handle the modopts yet

lambdageek committed:
[marshal] Add mono_marshal_get_native_func_wrapper_indirect

This will be used to add a wrapper around "calli sig" indirect calls where "sig" has
an unmanaged calling convention.

We need to do a GC transition before calling an unmanaged function from
managed. So this wrapper does it.

This is reusing much of the code implemented for
mono_marshal_get_native_func_wrapper_aot which is used for delegates.
Unfortunately that means that the function pointer is (pointlessly) boxed when
it is passed as the first argument.

lambdageek committed:
[jit] Use a wrapper for "calli unmanaged" instructions

If there's a calli with an unmanaged signature, invoke a wrapper that
does a transition to GC Safe mode and then do the call.

The wrapper is always inlined into the callee.

Because we're reusing much of the code of
mono_marshal_get_native_func_wrapper_aot, the function pointer first has to be
boxed before it's passed to the wrapper.  In theory we should be able to just
pass it directly and get much simpler code.  But that will require changing the
code in emit_native_wrapper_ilgen to get an unboxed function pointer arg

lambdageek commented:
@vargaz I'm not sure if I'm doing the stuff in `method-to-ir` in the best way.  Does this look ok?

lambdageek committed:
fixup don't emit GC transitions in runtime invoke wrapper

lambdageek committed:
fixup don't emit two wrappers on dynamic methods

lambdageek committed:
check that calli wrapper only gets blittable args

lambdageek committed:
add disabled debug printf

lambdageek committed:
negate logic for when to add an indirection wrapper

assume that if the callee is a wrapper, it doesn't need the transition, and
only add it to normal managed methods.

lambdageek commented:
Looks like it's passing all the tests if we only insert the GC transition for `calli` when it's not in a wrapper (ie it's a normal managed function).

There are a lot of different wrapper types, but I didn't go through all of them to see if it makes sense to omit the GC transition wrapper from all of them.  That means we're potentially going to be missing GC transitions in some wrappers.  But going the other way (only ignoring the native-to-managed wrapper and adding transitions in every other kind) caused a ton of crashes in mono/mono - things like cross-domain invokes, etc.

It would be nice to add a test that can verify that the pinvoked method is in GC Safe mode (using `mono_threads_assert_gc_safe_region` but maybe not only in checked build modes), but I haven't done it yet.

I'm going to rebase one more time to get rid of some debugging cruft but otherwise I think this is ready.

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3658,6 +3660,83 @@ mono_marshal_get_native_func_wrapper_aot (MonoClass *klass)
+	    return res;
+	
+	fprintf (stderr, "generating wrapper for signature %s\n", mono_signature_full_name (sig));
+	
```
Debug line.

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3658,6 +3660,83 @@ mono_marshal_get_native_func_wrapper_aot (MonoClass *klass)
+	}
+	for (int i = 0; i < sig->param_count; ++i)
+	{
+		MonoType *ty = sig->params[i];
```
Formatting.

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3658,6 +3660,83 @@ mono_marshal_get_native_func_wrapper_aot (MonoClass *klass)
+
+	if (!type_is_blittable (sig->ret)) {
+		g_assertf (type_is_blittable (sig->ret), "sig return type %s is not blittable\n", mono_type_full_name (sig->ret));
+	}
```
Is the if needed ?

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3658,6 +3660,83 @@ mono_marshal_get_native_func_wrapper_aot (MonoClass *klass)
+		MonoType *ty = sig->params[i];
+		if (!type_is_blittable (ty)) {
+			g_assertf (type_is_blittable (ty), "sig param %d (type %s) is not blittable\n", i, mono_type_full_name (ty));
+		}
```
Ditto.

vargaz commented on src/mono/mono/metadata/marshal.c:
```diff
@@ -3658,6 +3660,83 @@ mono_marshal_get_native_func_wrapper_aot (MonoClass *klass)
+	MonoMethodPInvoke *piinfo = NULL;
+	MonoMarshalSpec **mspecs = g_new0 (MonoMarshalSpec *, 1 + sig->param_count);
+        gpointer func = NULL;
+	gboolean check_exceptions = FALSE; /* FIXME: don't expect native to throw */
```
Formatting.

vargaz commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7065,6 +7065,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+					MonoMethod *wrapper = mono_marshal_get_native_func_wrapper_indirect (method->klass, fsig, cfg->compile_aot);
+					/* box the indirect function ptr */
+					/* FIXME: it would be nice if the wrapper didn't require a boxed argument */
+
```
Why does this need to be boxed ?

lambdageek replied:
I'm not sure. I'm reusing the IL generation in `emit_native_wrapper_ilgen` - and when it has to do an indirect call it expects the function to be boxed for some reason (I think it's used for delegates and it's more uniform if the first arg is always boxed.  but I'm not sure).

I didn't want to change it in this PR until I saw green tests, but it probably wouldn't be too hard to change it

vargaz replied:
This is because of the 'func_param' argument to emit_native_wrapper_ilgen which is used by delegate invokes. It probably needs another param which says that it receives an unboxed value.

vargaz commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7065,6 +7065,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+					if (inline_wrapper) {
+						int costs = inline_method (cfg, wrapper, fsig, sp, ip, cfg->real_offset, TRUE);
+						CHECK_CFG_EXCEPTION;
+						g_assert (costs > 0);
```
So because the wrapper is always inlined in the AOT case, the AOT code doesn't need to handle this kind of wrapper ?

lambdageek replied:
I think that's right.  But to be honest I'm just copying other uses of `inline_method` and `mono_emit_method_call` in method-to-ir.

vargaz commented on src/mono/mono/mini/method-to-ir.c:
```diff
@@ -7065,6 +7065,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
+						inline_costs += costs;
+						/* FIXME: calli_end expects ins to be the result of the call. */
+						ins = sp[0]; /*FIXME: is that the right instruction?*/
+					} else {
```
Yes, other cases do:
				if (!MONO_TYPE_IS_VOID (fsig->ret))
					/* *sp is already set by inline_method */
					ins = *sp;

lambdageek committed:
Add MonoNativeWrapperFlags arg to emit_native_wrapper

Replace the 4 boolean args by a flags arg.

Also add a new EMIT_NATIVE_WRAPPER_FUNC_PARAM_UNBOXED and use it in
mono_marshal_get_native_func_wrapper_indirect.  The new flag has no effect yet.

lambdageek committed:
Address review feedback

lambdageek committed:
Use an unboxed func ptr mono_marshal_get_native_func_wrapper_indirect

Add support for the EMIT_NATIVE_WRAPPER_FUNC_PARAM_UNBOXED flag to mono_marshal_emit_native_wrapper

lambdageek commented:
Addressed all the PR feedback, and added code to use an unboxed fnptr arg.  Tested on a HelloWorld sample locally that it emits the correct code.

vargaz commented on src/mono/mono/metadata/marshal-ilgen.c:
```diff
@@ -1995,13 +1995,20 @@ gc_safe_transition_builder_cleanup (GCSafeTransitionBuilder *builder)
+	gboolean check_exceptions = (flags & EMIT_NATIVE_WRAPPER_CHECK_EXCEPTIONS) != 0;
+	gboolean func_param = (flags & EMIT_NATIVE_WRAPPER_FUNC_PARAM) != 0;
+	/* TODO: make use of func_param_unboxed */
+	gboolean func_param_unboxed = (flags & EMIT_NATIVE_WRAPPER_FUNC_PARAM_UNBOXED) != 0;
```
TODO no longer needed.
